keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java

```diff
@@ -0,0 +1,53 @@
+package io.improbable.keanu.vertices.dbl.nonprobabilistic.operators.unary;
+
+import io.improbable.keanu.tensor.TensorShape;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.DualNumber;
+
+import java.util.Map;
+
+public class ReshapeVertex extends DoubleUnaryOpVertex {
+
+    private int[] proposedShape;
+
+    public ReshapeVertex(DoubleVertex inputVertex, int... proposedShape) {
+        super(inputVertex.getShape(), inputVertex);
+        this.proposedShape = proposedShape;
+    }
+
+    /**
+     * Returns the supplied vertex with a new shape of the same length
+     */
+    @Override
+    protected DoubleTensor op(DoubleTensor a) {
+        return a.reshape(proposedShape);
+    }
+
+    @Override
+    protected DualNumber calculateDualNumber(Map<Vertex, DualNumber> dualNumbers) {
+        Map<Long, DoubleTensor> partialDerivatives = dualNumbers.get(inputVertex).getPartialDerivatives().asMap();
```

Comment:
This reshaping of the dual numbers should probable happen in the DualNumbers class. There should be a new DualNumber for this vertex and the DualNumber for inputVertex shouldn't be changed or removed.

```
DualNumber reshapedDualNumber = dualNumbers.get(inputVertex).reshape(proposedShape)
dualNumbers.put(getId(), reshapedDualNumber);
```

Reply:
Nice, looks much cleaner. changed.